pkg/broker/config/key.go

```diff
@@ -30,51 +30,98 @@ import (
+	for k, v := range CellTenantType_value {
+		t := CellTenantType(v)
+		lowerCase := strings.ToLower(k)
+		cellTenantTypeFromLowerCase[lowerCase] = t
+		cellTenantTypeToLowerCase[t] = lowerCase
+	}
+}
+
+// CellTenantKey uniquely identifies a single CellTenant, at a given point in time.
+type CellTenantKey struct {
+	cellTenantType CellTenantType
+	namespace      string
+	name           string
 }
 
 // String creates a human readable version of this key. It is for debug purposes only. It is free to
 // change at any time.
-func (k *BrokerKey) String() string {
+func (k *CellTenantKey) String() string {
 	// Note that this is explicitly different than the PersistenceString, so that we don't
 	// accidentally use String(), rather than PersistenceString().
-	return k.namespace + "//" + k.name
+	return fmt.Sprintf("%s:%s//%s", k.cellTenantType, k.namespace, k.name)
 }
 
 // PersistenceString is the string that is persisted as the key for this Broker in the protobuf. It
 // is stable and can only change if all existing usage locations are made backwards compatible,
 // supporting _both_ the old and the new format, for at least one release.
-func (k *BrokerKey) PersistenceString() string {
-	return k.namespace + "/" + k.name
+func (k *CellTenantKey) PersistenceString() string {
+	// TODO It is safe to remove this after release-0.22, decide whether we want to leave Broker
+	// special-cased or not.
+	if k.cellTenantType == CellTenantType_BROKER {
+		// For backwards compatibility from when the only type was Broker, Brokers do not embed
+		// their type into the string.
+		return k.namespace + "/" + k.name
+	}
+	return fmt.Sprintf("%s/%s/%s", cellTenantTypeToLowerCase[k.cellTenantType], k.namespace, k.name)
 }
 
-func BrokerKeyFromPersistenceString(s string) (*BrokerKey, error) {
+func CellTenantKeyFromPersistenceString(s string) (*CellTenantKey, error) {
 	pieces := strings.Split(s, "/")
-	if len(pieces) != 3 {
-		return nil, fmt.Errorf("malformed request path; expect format '/<ns>/<broker>', actually %q", s)
+	if len(pieces) <= 2 || len(pieces) >= 5 {
+		return nil, fmt.Errorf(
+			"malformed request path; expect format '/<ns>/<broker>' or '/<type>/<ns>/<broker>', actually %q", s)
 	}
-	// Broker's persistence strings are in the form "/<ns>/<brokerName>".
-	blank, ns, brokerName := pieces[0], pieces[1], pieces[2]
+	var blank, ns, brokerName string
+	var t CellTenantType
+	if len(pieces) == 3 {
+		// Broker's persistence strings are in the form "/<ns>/<brokerName>".
+		blank, ns, brokerName = pieces[0], pieces[1], pieces[2]
+		t = CellTenantType_BROKER
+	} else {
+		// len(pieces) must be 4, so this is the standard form of the persistence string,
+		// '/<type>/<ns>/<name>'.
+		var ts string
+		blank, ts, ns, brokerName = pieces[0], pieces[1], pieces[2], pieces[3]
+		if ctt, err := validateCellTenantTypeFromString(ts); err != nil {
+			return nil, err
+		} else {
+			t = ctt
+		}
```

Comment:
I thought about it, but have a strong dislike for when `:=` and `=` would cause issues. In this case, the compiler would still catch the `=` error, as `t` would be unused, but even so, I prefer it this more explicit way. What do you think?